codegen-serde/src/main/kotlin/software/amazon/smithy/rust/codegen/serde/SerializeImplGenerator.kt

```diff
@@ -0,0 +1,392 @@
+                )
+            }
+        }
+
+    /**
+     * Serialize a type that already implements `Serialize` directly via `value.serialize(serializer)`
+     */
+    private fun directSerde(shape: Shape): RuntimeType {
+        return RuntimeType.forInlineFun(codegenContext.symbolProvider.toSymbol(shape).rustType().toString(), SerdeModule) {
+            implSerializeConfigured(codegenContext.symbolProvider.toSymbol(shape)) {
+                val baseValue =
+                    writable { rust("self.value") }.letIf(shape.isStringShape) { it.plus(writable(".as_str()")) }
+                rustTemplate("#{base}.serialize(serializer)", "base" to baseValue)
+            }
+        }
+    }
+
+    /**
+     * Serialize a shape by first generating a wrapper struct:
+     * ```rust
+     * struct WrapperType<'a>(&'a Type);
+     * ```
+     *
+     * Then implementing `Serialize` for `ConfigurableSerdeRef<'a WrapperType>`
+     *
+     * This exists to allow differing implementations for same-shaped-rust types. For example, consider
+     * the following Smithy model:
+     *
+     * ```smithy
+     * list SensitiveList {
+     *   member: SensitiveString
+     * }
+     *
+     * list StringList {
+     *   member: String
+     * }
+     * ```
+     *
+     * These both are `Vec<String>` but must be serialized differently.
+     */
+    private fun serializeWithWrapper(
+        shape: Shape,
+        body: (Writable) -> Writable,
+    ): RuntimeType {
+        val name =
+            "Serializer" +
+                codegenContext.symbolProvider.shapeFunctionName(codegenContext.serviceShape, shape)
+                    .toPascalCase()
+        // awkward hack to recover the symbol referring to the type
+        val type = SerdeModule.toType().resolve(name).toSymbol()
+        val base = writable { rust("self.value.0") }
+        val serialization =
+            implSerializeConfiguredWrapper(type) {
+                body(base)(this)
+            }
+        val wrapperStruct =
+            RuntimeType.forInlineFun(name, SerdeModule) {
+                rustTemplate(
+                    """
+                    struct $name<'a>(&'a #{Shape});
+                    #{serializer}
+                    """,
+                    "Shape" to codegenContext.symbolProvider.toSymbol(shape),
+                    "serializer" to serialization,
+                )
+            }
+        return wrapperStruct
+    }
+
+    /**
+     * Serialize the field of a structure, union, list or map.
+     *
+     * All actual serialization MUST go through this path as it handles applying the `Sensitive` wrapper.
+     */
+    private fun serializeMember(
+        shape: MemberShape,
+        memberRef: String,
+    ): Writable {
+        val target = codegenContext.model.expectShape(shape.target)
+        return writable {
+            serializerFn(target) {
+                rust("&$memberRef")
+            }.plus { rust(".serialize_ref(&self.settings)") }(this)
```

Comment:
Hmm Clippy should catch this, there's no need for `&` here. Which makes me think that we're not running `cargo clippy --all-features` in CI.

Reply:
fixed clippy lints in the generated code